### PR TITLE
Port to using node-cmake

### PR DESCRIPTION
**Pull request by TheMarex**
Depends on https://github.com/Project-OSRM/osrm-backend/pull/3367

- [x] Rebase + squash
- [x] Final review by @daniel-j-h

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+if(NOT CMAKE_BUILD_TYPE)
+  message(STATUS "No build type specified, defaulting to Release")
+  set(CMAKE_BUILD_TYPE Release)
+endif()
+
+option(BUILD_LIBOSRM "Download and build own libsorm version" OFF)
+
+set(OSRM_BINARIES "")
+if(BUILD_LIBOSRM)
+  set(OSRM_INSTALL_DIR "/tmp/osrm-backend")
```
Comment (daniel-j-h) [inline CMakeLists.txt]: Hrm why do we put it there? Asking since it may clash with existing osrm-backend dirs there.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    file(GLOB TBBGlob ${libpath}/*.*)
+    foreach(filepath ${TBBGlob})
+      get_filename_component(filename ${filepath} NAME)
+      set(targetfile "${CMAKE_SOURCE_DIR}/lib/binding/${filename}")
+      add_custom_command(OUTPUT ${targetfile}
+                         COMMAND ${CMAKE_COMMAND} -E copy ${filepath} ${CMAKE_SOURCE_DIR}/lib/binding/
+                         DEPENDS ${filepath})
+      list(APPEND OSRM_BINARIES "${targetfile}")
+    endforeach()
+  endforeach()
```
Comment (daniel-j-h) [inline CMakeLists.txt]: This copies over the tbb shared libs which libosrm set an rpath for, right? ([ref.](https://github.com/Project-OSRM/osrm-backend/blob/aaf42a1cafba62f291be132b7027a635ed5b76c5/CMakeLists.txt#L376-L381)).
Which means we have to overwrite the old rpath with `chrpath` fixing it according to our dirs (?)

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  # LibOSRM_DEPENDENT_LIBRARIES
+  # LibOSRM_CXXFLAGS
+  find_package(LibOSRM REQUIRED)
+endif()
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/node_modules/node-cmake")
+
+set(NodeJS_CXX_STANDARD 14 CACHE INTERNAL "Use C++14" FORCE)
+set(NodeJS_DOWNLOAD ON CACHE INTERNAL "Download node.js sources" FORCE)
```
Comment (daniel-j-h) [inline CMakeLists.txt]: This is what I stumbled upon a couple of month ago, too. Why do we have to make a HTTP request for C++ header when the NodeJs headers are available locally or - aren't they always available? For example node-cmake could not detect them installed via nvm at `~/.nvm/versions/node/vx.x.x/include/` but that looks like a node-cmake issue?

File: README.md [hunk 4]
```diff
@@ -121,63 +121,36 @@ make && make test
...
-npm install --build-from-source
-# build the osrm-backend test data
-make -C ../osrm-backend/test/data
-export OSRM_DATA_PATH=../osrm-backend/test/data
-npm test
-```
+    mkdir build
+    cd build
+    cmake ..
+    make clean
```
Comment (daniel-j-h) [inline README.md]: Why make clean as first target?

File: CMakeLists.txt (hunk 1 above)
Comment (TheMarex) [inline CMakeLists.txt]: Actually this is not needed anymore. Good catch.

File: README.md (hunk 4 above)
Comment (TheMarex) [inline README.md]: To remove any possible previous installation that could conflict (for example downloaded binaries).

File: CMakeLists.txt (hunk 2 above)
Comment (DerDakon) [inline CMakeLists.txt]: usually CMake will set a RPATH for every shared object (i.e. also executables) so they will find their dependencies from the current build tree, e.g. if you build both a shared library and an executable using it. These RPATH entries will be changed on "make install" to point to the final destination.

File: README.md (hunk 4 above)
Comment (DerDakon) [inline README.md]: Just "rm -rf" the build directory before then, this has the advantage that it also works e.g. when switching the CMake version.

File: Makefile [hunk 5]
```diff
@@ -1,3 +1,15 @@
+# If JOBS environment variable not defined, default to the number of CPUs available
+ifeq ($(JOBS),)
+JOBS:=1
+OS:=$(shell uname -s)
+ifeq ($(OS),Linux)
+  JOBS:=$(shell grep -c ^processor /proc/cpuinfo)
```
Comment (DerDakon) [inline Makefile]: Try "getconf _NPROCESSORS_ONLN"

Comment (danpat) [inline Makefile]: Well, this clause is already checking for `Linux`, so I feel confident that `/proc/cpuinfo` will exist.  To be honest, I just copy/pasted this from somewhere else.  I don't feel strongly either way, just as long as it works on Travis :-)

Comment (DerDakon) [inline Makefile]: If you want somewhere to copy&paste from, try [ProcessorCount.cmake](https://github.com/Kitware/CMake/blob/master/Modules/ProcessorCount.cmake), where we made that work for all platforms we could get access to.

File: CMakeLists.txt (hunk 3 above)
Comment (springmeyer) [inline CMakeLists.txt]: > Why do we have to make a HTTP request for C++ header when the NodeJs headers are available locally or - aren't they always available?

FWIW node-gyp does this too: it downloads the right headers dynamically for a given node version. I think the idea is to not have to depend on the existing node installation (and brittle searching for its include directories) and to instead download the right package. The way that node-gyp allows using the headers for a given local node installation is through the `--nodedir` flag (https://github.com/nodejs/node-gyp/blob/9c8d275526c770f4e17955b2d8121b635b7ad187/README.md#command-options)

**Comment by springmeyer** [comment]
I noticed that the install_name (aka id) for the `osrm.node` is wrong. This is harmless since nothing links to `osrm.node` but still should likely be fixed:

```
$ ldd lib/binding/osrm.node 
lib/binding/osrm.node:
	@rpath/node-osrm.node (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libtbb.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```

It should be `@rpath/osrm.node` instead of `@rpath/node-osrm.node`.

**Pull request merged**